### PR TITLE
Use a small cookie to do cache poisoning detection.

### PR DESCRIPTION
**Pull request by ajacksified**
:eyeglasses: @JordanMilne

### CODE REVIEW

File: assets/js/client.es6.js [hunk 1]
```diff
@@ -45,6 +46,86 @@ function onLoad(fn) {
...
+    });
+  }
+}
+
+function checkCanaryCookie(config) {
+  var canary = cookies.get(CANARY_COOKIE_NAME);
+
+  // There was a canary from the page, which sould only appear for logged-in
+  // users. So check it against our canary to make sure they match.
+  if (config.canary && config.canary !== canary) {
```
Comment (JordanMilne) [inline assets/js/client.es6.js]: need an `&& canary` here too, the canary cookie might not be set yet.

File: assets/js/client.es6.js [hunk 2]
```diff
@@ -45,6 +46,86 @@ function onLoad(fn) {
...
+}
+
+var csrfRegex = /csrf-token" content="([^"]+)/;
+
+function logCachePoisoning(config) {
+  superagent
+    .get(window.location.href)
+    .end(function (err, res) {
+      var details = {
+        poisoner_name: config.user.name,
```
Comment (JordanMilne) [inline assets/js/client.es6.js]: need `victim_canary` here too, the server double checks that they're actually different

File: assets/js/client.es6.js [hunk 3]
```diff
@@ -45,6 +46,86 @@ function onLoad(fn) {
...
+        .send(details)
+        .set({
+          'X-Loggit': true,
+        })
+        .end(function(){});
+    });
+}
+
+function checkCachePoisoning(config) {
+  if (!checkCanaryCookie(config)) {
```
Comment (JordanMilne) [inline assets/js/client.es6.js]: Does this make sure that this is a logged-in page? We don't care if loggedout pages end up in caches

File: src/routes.jsx [hunk 4]
```diff
@@ -218,6 +218,7 @@ function routes(app) {
       origin: app.getConfig('origin'),
       propertyId: app.getConfig('googleAnalyticsId'),
       userAgent: ctx.userAgent,
+      canary: ctx.canary,
```
Comment (JordanMilne) [inline src/routes.jsx]: Same as above, does this only put this on the page if logged in?

File: src/routes.jsx [hunk 5]
```diff
@@ -218,6 +218,7 @@ function routes(app) {
       origin: app.getConfig('origin'),
       propertyId: app.getConfig('googleAnalyticsId'),
       userAgent: ctx.userAgent,
```
Comment (JordanMilne) [inline src/routes.jsx]: You'll also need a `cachePolicy` attr to differentiate between different cache policies used on the same app. 'loggedin_mweb' if loggedin and 'loggedout_mweb' if loggedout are what I'm relying on for now.

File: src/routes.jsx (hunk 4 above)
Comment (ajacksified) [inline src/routes.jsx]: with https://github.com/ajacksified/reddit-mobile/commit/ec6c8407e228f14a03b4f4b735e3a28abb041b53 - yes.

File: assets/js/client.es6.js (hunk 3 above)
Comment (ajacksified) [inline assets/js/client.es6.js]: `config.canary` will be empty if you're logged in, so it won't run the check (`if (canary && config.canary && config.canary !== canary)` will fail)

File: assets/js/client.es6.js [hunk 6]
```diff
@@ -45,6 +46,87 @@ function onLoad(fn) {
...
+
+var csrfRegex = /csrf-token" content="([^"]+)/;
+
+function logCachePoisoning(config) {
+  superagent
+    .get(window.location.href)
+    .end(function (err, res) {
+      var details = {
+        poisoner_name: config.user.name,
+        poisoner_id: config.user.id,
```
Comment (JordanMilne) [inline assets/js/client.es6.js]: You'll need to MAC some of this data, include it in the page, and then send it off with any reports, see  how `make_poisoning_report_mac` is used in r2

File: src/routes.jsx (hunk 5 above)
Comment (JordanMilne) [inline src/routes.jsx]: Oh, and also a `renderTime` property with a JS timestamp of when the page was rendered. That way we know how long the page has been sitting around in the cache.

**Comment by JordanMilne** [comment]
Definitely a lot cleaner than what I was doing! Just a couple things :nail_care:

**Comment by ajacksified** [comment]
:haircut:

File: src/views/layouts/DefaultLayout.jsx [hunk 7]
```diff
@@ -1,8 +1,38 @@
...
+  hmac.setEncoding('hex');
+  hmac.write(data);
+  hmac.end();
+
+  return hmac.read();
+}
+
+function generateCanaryMac(props, secret) {
+  var macData = [
+    props.user.name,
```
Comment (JordanMilne) [inline src/views/layouts/DefaultLayout.jsx]: the canary itself should be the first thing in `macData`

**Comment by ajacksified** [comment]
:haircut:

**Comment by JordanMilne** [comment]
:fish: lgtm, but deploy's blocked until I fix the CORS issues so you can send the report :sleeping:

**Comment by JordanMilne** [comment]
We _should_ be good to go ahead with this now

**Pull request closed without merging**